DataFormats/PortableTestObjects/src/alpaka/classes_serial_def.xml

```diff
@@ -5,7 +5,7 @@
     targetClass="portabletest::TestHostCollection"
     version="[1-]"
     source="portabletest::TestSoA layout_;"
-    target="buffer_"
+    target="buffer_,layout_,view_"
```

Comment:
I'm happy if we get rid of memory leaks - but why do we have to specify `layout_` and `view_` in the first place ?
They are declared as transient-only, should ROOT leave them alone ?

Reply:
We are redundant. I would say he semantic here is that the streamer will initialize those variables (which is true), but being transient-only, ROOT will not touch them either. We could decide on both convention (minimum information or exhaustively declaring the member variables we initialize (all, in fact).

Reply:
I'm a bit confused, `layout_` is the persisted member, no?

Reply:
> * `layout_` is persistent; however we want it to use our user-defined memory, not the one provided by ROOT; is this why we need to list it as a "target" ?

Correct.

> * `view_` is transient, so I would expect ROOT not to allocate any memory for it; why do we need to list it as a "target" ?

The `view_` was indeed not part of the minimal fix suggestion. @pcanal, does including a transient `view_` member in the `target` list have impact on any kind from ROOT side?

Reply:
I would say `buffer_` and `view_` have similar statuses and would prefer have both either listed or not. As they are declared transient in the code, it is most likely redundant (TBC by @pcanal) to list them here, but that gives a complete description of the streamer's behavior.

Reply:
If/since `view_` is also affected by this rule, it make sense to list it as part of the target.  (This let the system knows that if `view_` is directly ask for, it should run the rule.

Reply:
I'm afraid I'm still confused :-(

Without this change, reflex generates this code for the class dictionary:
```c++
    // Schema evolution read functions
    static void read_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_0( char* target, TVirtualObject *oldObj )
    {
       //--- Automatically generated variables ---
       struct PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile {
          portabletest::TestSoA &layout_;
          PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile(portabletest::TestSoA &onfile_layout_ ): layout_(onfile_layout_) {}
       };
       static Long_t offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_ = oldObj->GetClass()->GetDataMemberOffset("layout_");
       char *onfile_add = (char*)oldObj->GetObject();
       PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile onfile(
          *(portabletest::TestSoA*)(onfile_add+offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_) );

       static TClassRef cls("portabletest::TestHostCollection");
       static Long_t offset_buffer_ = cls->GetDataMemberOffset("buffer_");
       optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>& buffer_ = *(optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>*)(target+offset_buffer_);
       portabletest::TestHostCollection* newObj = (portabletest::TestHostCollection*)target;

       // Supress warning message.
       (void)oldObj;
       (void)newObj;

       //--- User's code ---
       portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
    }
```

With this change, the function becomes
```c++
    // Schema evolution read functions
    static void read_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_0( char* target, TVirtualObject *oldObj )
    {
       //--- Automatically generated variables ---
       struct PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile {
          portabletest::TestSoA &layout_;
          PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile(portabletest::TestSoA &onfile_layout_ ): layout_(onfile_layout_) {}
       };
       static Long_t offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_ = oldObj->GetClass()->GetDataMemberOffset("layout_");
       char *onfile_add = (char*)oldObj->GetObject();
       PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_Onfile onfile(
          *(portabletest::TestSoA*)(onfile_add+offset_Onfile_PortableHostCollectionlEportabletestcLcLTestSoALayoutlE128cOfalsegRsPgR_layout_) );

       static TClassRef cls("portabletest::TestHostCollection");
       static Long_t offset_buffer_ = cls->GetDataMemberOffset("buffer_");
       optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>& buffer_ = *(optional<PortableHostCollection<portabletest::TestSoALayout<128,false> >::Buffer>*)(target+offset_buffer_);
       static Long_t offset_layout_ = cls->GetDataMemberOffset("layout_");
       PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout& layout_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout*)(target+offset_layout_);
       static Long_t offset_view_ = cls->GetDataMemberOffset("view_");
       PortableHostCollection<portabletest::TestSoALayout<128,false> >::View& view_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::View*)(target+offset_view_);
       portabletest::TestHostCollection* newObj = (portabletest::TestHostCollection*)target;

       // Supress warning message.
       (void)oldObj;
       (void)newObj;

       //--- User's code ---
       portabletest::TestHostCollection::ROOTReadStreamer(newObj, onfile.layout_);
    }
```

The additional code is
```c++
      static Long_t offset_layout_ = cls->GetDataMemberOffset("layout_");
      PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout& layout_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::Layout*)(target+offset_layout_);
      static Long_t offset_view_ = cls->GetDataMemberOffset("view_");
      PortableHostCollection<portabletest::TestSoALayout<128,false> >::View& view_ = *(PortableHostCollection<portabletest::TestSoALayout<128,false> >::View*)(target+offset_view_);
```
which simply assigns pointers into the red buffer to references to `layout_` and `view_`.

But I don't see anything anywhere about not allocating memory for them, or deleting them.
In fact, these references are not used at all.

Why should this make any difference ?

Reply:
The difference is not in the generated code.  (The generator does not parse the user `code` for the rule and thus set a series of pointer `just in case` for old the dependencies).

The difference comes from how the `TStreamerInfo` infrastructure is handling the dependencies.

With a struct like:
```
struct Example {
   InnerClass x;
   Transient z;  //!
};
```
A rule listing:
```
source="InnerClass x" target="z"
```
'just' say that `x read from the file` is the input for z.  Consequently it *also* need to read `x` (a second time) to update the in memory x.

A rule listing:
```
source="InnerClass x" target="x,z"
```
says that rule will update `x and z` from the input and there is no need to read it a second time.

Another example:

```
struct Example2
{
     float x;
     float y;
     float z;
};
```
If on file with have `x==2` and `y==5`
with the rules
```
source="float x" target="x" code="{ x = onfile.x * 2; }"
source="float x" target="z" code="{ z =  onfile.x; };
```
after reading we have `x==4` and `z==2`

In other word the `onfile.MemberName` and `MemberName` are independent and unless it was indicated that `MemberName` will be explicitly set by a rule (i.e. is a target) then we have to read it a second time if it is use as input of a rule (because we need to give the rule the not updated version).

Reply:
I see, thanks for the explanation.